luma/core/bitmap_font.py

```diff
@@ -0,0 +1,632 @@
+            the source_font will be transferred.
+        :type characters: str
+        :param force: If set, the source_font can overwrite values that already
+            exists within this font.  Default is False.
+        :type force: bool
+        """
+
+        if characters:
+            for c in characters:
+                if ord(c) in self.mappings and not force:
+                    continue
+                m = source_font.lookup(ord(c))
+                if m is not None:
+                    v = source_font.metrics[m]
+                else:
+                    raise ValueError('{0} is not a valid character within the source font'.format(c))
+                self.metrics.append(v)
+                self.mappings[ord(c)] = len(self.metrics) - 1
+        else:
+            # Copy source values into destination but don't overwrite existing characters unless force set
+            for k, v in source_font.mappings.items():
+                if k in self.mappings and not force:
+                    continue
+                self.metrics.append(source_font.metrics[v])
+                self.mappings[k] = len(self.metrics) - 1
+
+        # Recompute font size metrics
+        self._calculate_font_size()
+
+
+def load(filename):
+    """
+    Load a LUMA.CORE.BITMAP_FONT file.  This function creates a
+    luma.core.bitmap_font object from the given LUMA.CORE.BITMAP_FONT file, and
+    returns the corresponding font object.
+
+    :param filename: Filename of font file.
+    :type filename: str
+    :return: A luma.core.bitmap_font object.
+    :exception OSError: If the file could not be read.
+    :exception SyntaxError: If the file does not contain the expected data
+    """
+    f = bitmap_font()
+    f.load(filename)
+    return f
+
+
+def loads(data):
+    """
+    Load a luma.core.bitmap_font from a string of serialized data.  This function
+    creates a luma.core.bitmap object from serialized data produced from the
+    dumps method and returns the corresponding font object.
+
+    :param data: Serialized luma.core.bitmap_font data.
+    :type data: str
+    :return: A luma.core.bitmap_font object.
+    :exception ValueError: If the data does not a valid luma.core.bitmap_font
+    """
+    f = bitmap_font()
+    f.loads(data)
+    return f
+
+
+def load_pil_font(filename, mappings=None):
+    """
+    Load a PIL font file.  This function creates a luma.core.bitmap_font object
+    from the given PIL bitmap font file, and returns the corresponding font object.
+
+    :param filename: Filename of font file.
+    :type filename: str
+    :param mappings: a dictionary of unicode to value pairs (optional)
+    :type mappings: dict
+    :return: A font object.
+    :exception OSError: If the file could not be read.
+    :exception SyntaxError: If the file does not contain the expected data
+    """
+    f = bitmap_font()
+    f.load_pil_font(filename, mappings)
+    return f
```

Comment:
is this file ever closed?

Reply:
The file is closed after producing the resulting bitmap_font object.